Triggering IConnectionLifetimeNotificationFeature on GoAway

ladeak opened the pull request:
Triggering Connection Lifetime Notification Feature's cancellation on GoAway

On GoAway frames triggering `IConnectionLifetimeNotificationFeature`'s cancellation for gRPC and SignalR to initiate teardown of long running requests.

## Description

For Kestrel h2 and h3 triggering `IConnectionLifetimeNotificationFeature` feature's *RequestClose()* method on GoAway frames, so that frameworks like SignalR or gRPC may proactively start closing long running request.
h3 control stream (which is the only stream that handles GoAway) also calls *StopProcessingNextRequest()* method otherwise `HttpConnection` (subscribed to `IConnectionLifetimeNotificationFeature`) would send GoAway's from server. h2 already does the same.

Adding new tests: `Http3InMemory` now implements *RequestClose()* method and `Http2TestBase` mimics sharing the same feature collection on ConnectionFeatures and by the Connection's Features properties.


Fixes #42057

halter73 commented on src/Servers/Kestrel/shared/test/Http3/Http3InMemory.cs:
```diff
@@ -1073,7 +1073,7 @@ public void OnHeartbeat(Action<object> action, object state)
     public void RequestClose()
     {
-        throw new NotImplementedException();
+        ConnectionClosingCts.Cancel();
```
Were you able to test this with a real HTTP/3 connection? It looks like it should work, but I want to be sure.

ladeak replied:
Not yet. I could not figure out how to make HttpClient send a GoAway. I am looking at building something to test it, but it might take some time. I could only test through the existing tests for now.

ladeak replied:
I am testing this with .NET6 app -- h3 --> YARP --h3--> Sample ASP.NET Core webapi.

The ASP.NET Core webapi awaits Task.Delay(...) before writing the response. Until this I shut down the YARP proxy.

I don't receive a GOAWAY on the webapi.
I don't see YARP would send a GOAWAY to the webapi.

I see a GOAWAY request for the. Here are some logs:

ASP.NET Core webapi:

```
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[1]
      Connection id "0HMK61V4M5U2C" accepted.
dbug: Microsoft.AspNetCore.Server.Kestrel.Connections[39]
      Connection id "0HMK61V4M5U2C" accepted.
dbug: Microsoft.AspNetCore.Server.Kestrel.Connections[1]
      Connection id "0HMK61V4M5U2C" started.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[3]
      Stream id "0HMK61V4M5U2C:00000003" type Unidirectional connected.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[2]
      Stream id "0HMK61V4M5U2C:00000002" type Unidirectional accepted.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[2]
      Stream id "0HMK61V4M5U2C:00000000" type Bidirectional accepted.
trce: Microsoft.AspNetCore.Server.Kestrel.Http3[46]
      Connection id "0HMK61V4M5U2C" received SETTINGS frame for stream ID 2 with length 5.
trce: Microsoft.AspNetCore.Server.Kestrel.Http3[46]
      Connection id "0HMK61V4M5U2C" received HEADERS frame for stream ID 0 with length 194.
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/3 GET https://localhost:5001/ - -
trce: Microsoft.AspNetCore.Server.Kestrel.Http3[47]
      Connection id "0HMK61V4M5U2C" sending HEADERS frame for stream ID 0 with length 44.
trce: Microsoft.AspNetCore.Server.Kestrel.Http3[47]
      Connection id "0HMK61V4M5U2C" sending DATA frame for stream ID 0 with length 19.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[10]
      Stream id "0HMK61V4M5U2C:00000000" shutting down writes because: "The QUIC transport's send loop completed gracefully.".
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/3 GET https://localhost:5001/ - - - 200 - - 11685.0355ms
dbug: Microsoft.AspNetCore.Server.Kestrel[25]
      Connection id "0HMK61V4M5U2C", Request id "0HMK61V4M5U2C:00000000": started reading request body.
dbug: Microsoft.AspNetCore.Server.Kestrel[26]
      Connection id "0HMK61V4M5U2C", Request id "0HMK61V4M5U2C:00000000": done reading request body.
trce: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[16]
      Stream id "0HMK61V4M5U2C:00000000" pooled for reuse.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[7]
      Stream id "0HMK61V4M5U2C:00000002" unexpected error.
      System.Net.Quic.QuicException: The connection timed out from inactivity.
         at System.Net.Quic.ResettableValueTaskSource.TryComplete(Exception exception, Boolean final)
         at System.Net.Quic.QuicStream.HandleEventShutdownComplete(_SHUTDOWN_COMPLETE_e__Struct& data)
         at System.Net.Quic.QuicStream.HandleStreamEvent(QUIC_STREAM_EVENT& streamEvent)
         at System.Net.Quic.QuicStream.NativeCallback(QUIC_HANDLE* connection, Void* context, QUIC_STREAM_EVENT* streamEvent)
      --- End of stack trace from previous location ---
         at System.Net.Quic.ResettableValueTaskSource.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
         at System.Net.Quic.QuicStream.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
         at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Internal.QuicStreamContext.DoReceiveAsync() in C:\Users\ladea\source\repos\aspnetcore\src\Servers\Kestrel\Transport.Quic\src\Internal\QuicStreamContext.cs:line 214
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[7]
      Stream id "0HMK61V4M5U2C:00000003" unexpected error.
      System.Net.Quic.QuicException: The connection timed out from inactivity.
         at System.Net.Quic.ResettableValueTaskSource.TryComplete(Exception exception, Boolean final)
         at System.Net.Quic.QuicStream.HandleEventShutdownComplete(_SHUTDOWN_COMPLETE_e__Struct& data)
         at System.Net.Quic.QuicStream.HandleStreamEvent(QUIC_STREAM_EVENT& streamEvent)
         at System.Net.Quic.QuicStream.NativeCallback(QUIC_HANDLE* connection, Void* context, QUIC_STREAM_EVENT* streamEvent)
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Internal.QuicStreamContext.WaitForWritesClosedAsync() in C:\Users\ladea\source\repos\aspnetcore\src\Servers\Kestrel\Transport.Quic\src\Internal\QuicStreamContext.cs:line 188
         at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Internal.QuicStreamContext.DoSendAsync() in C:\Users\ladea\source\repos\aspnetcore\src\Servers\Kestrel\Transport.Quic\src\Internal\QuicStreamContext.cs:line 394
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[10]
      Stream id "0HMK61V4M5U2C:00000003" shutting down writes because: "The connection timed out from inactivity.".
dbug: Microsoft.AspNetCore.Server.Kestrel.Http3[45]
      Trace id "0HMK61V4M5U2C": HTTP/3 stream error "H3_INTERNAL_ERROR". An abort is being sent to the stream.
      Microsoft.AspNetCore.Connections.ConnectionAbortedException: HTTP_CLOSED_CRITICAL_STREAM
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[13]
      Stream id "0HMK61V4M5U2C:00000002" aborted by application with error code 258 because: "HTTP_CLOSED_CRITICAL_STREAM".
dbug: Microsoft.AspNetCore.Server.Kestrel.BadRequests[20]
      Connection id "0HMK61V4M5U2C" request processing ended abnormally.
      System.Net.Quic.QuicException: The connection timed out from inactivity.
         at System.Net.Quic.QuicConnection.HandleEventShutdownInitiatedByTransport(_SHUTDOWN_INITIATED_BY_TRANSPORT_e__Struct& data)
         at System.Net.Quic.QuicConnection.HandleConnectionEvent(QUIC_CONNECTION_EVENT& connectionEvent)
         at System.Net.Quic.QuicConnection.NativeCallback(QUIC_HANDLE* connection, Void* context, QUIC_CONNECTION_EVENT* connectionEvent)
      --- End of stack trace from previous location ---
         at System.Net.Quic.QuicConnection.AcceptInboundStreamAsync(CancellationToken cancellationToken)
         at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Internal.QuicConnectionContext.AcceptAsync(CancellationToken cancellationToken) in C:\Users\ladea\source\repos\aspnetcore\src\Servers\Kestrel\Transport.Quic\src\Internal\QuicConnectionContext.cs:line 96
         at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3.Http3Connection.ProcessRequestsAsync[TContext](IHttpApplication`1 application) in C:\Users\ladea\source\repos\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http3\Http3Connection.cs:line 311
dbug: Microsoft.AspNetCore.Server.Kestrel.Http3[53]
      Connection id "0HMK61V4M5U2C": GOAWAY stream ID 4.
trce: Microsoft.AspNetCore.Server.Kestrel.Http3[47]
      Connection id "0HMK61V4M5U2C" sending GOAWAY frame for stream ID 3 with length 1.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[6]
      Connection id "0HMK61V4M5U2C" aborted by application with error code -1 because: "The HTTP/3 connection faulted.".
dbug: Microsoft.AspNetCore.Server.Kestrel.Http3[44]
      Connection id "0HMK61V4M5U2C" is closed. The last processed stream ID was 0.
dbug: Microsoft.AspNetCore.Server.Kestrel.Connections[2]
      Connection id "0HMK61V4M5U2C" stopped.
```

YARP:

```
      Hosting started
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[1]
      Connection id "0HMK61V4ICAUC" accepted.
dbug: Microsoft.AspNetCore.Server.Kestrel.Connections[39]
      Connection id "0HMK61V4ICAUC" accepted.
dbug: Microsoft.AspNetCore.Server.Kestrel.Connections[1]
      Connection id "0HMK61V4ICAUC" started.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[3]
      Stream id "0HMK61V4ICAUC:00000003" type Unidirectional connected.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[2]
      Stream id "0HMK61V4ICAUC:00000002" type Unidirectional accepted.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[2]
      Stream id "0HMK61V4ICAUC:00000000" type Bidirectional accepted.
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/3 GET https://localhost:7007/ - -
dbug: Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware[0]
      Wildcard detected, all requests with hosts will be allowed.
dbug: Microsoft.AspNetCore.Routing.Matching.DfaMatcher[1001]
      1 candidate(s) found for the request path '/'
dbug: Microsoft.AspNetCore.Routing.Matching.DfaMatcher[1005]
      Endpoint 'route1' with route pattern '{**catch-all}' is valid for the request path '/'
dbug: Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware[1]
      Request matched endpoint 'route1'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint 'route1'
info: Yarp.ReverseProxy.Forwarder.HttpForwarder[9]
      Proxying to https://localhost:5001/ HTTP/3 RequestVersionOrHigher no-streaming
dbug: Microsoft.AspNetCore.Server.Kestrel[25]
      Connection id "0HMK61V4ICAUC", Request id "0HMK61V4ICAUC:00000000": started reading request body.
dbug: Microsoft.AspNetCore.Server.Kestrel[26]
      Connection id "0HMK61V4ICAUC", Request id "0HMK61V4ICAUC:00000000": done reading request body.
info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...
dbug: Microsoft.Extensions.Hosting.Internal.Host[3]
      Hosting stopping
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[0]
      Listener has aborted with exception: Operation aborted.
dbug: Microsoft.AspNetCore.Server.Kestrel.Http3[43]
      Connection id "0HMK61V4ICAUC" is closing.
dbug: Microsoft.AspNetCore.Server.Kestrel.Http3[53]
      Connection id "0HMK61V4ICAUC": GOAWAY stream ID 4611686018427387903.
info: Yarp.ReverseProxy.Forwarder.HttpForwarder[56]
      Received HTTP/3.0 response 200.
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint 'route1'
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[10]
      Stream id "0HMK61V4ICAUC:00000000" shutting down writes because: "The QUIC transport's send loop completed gracefully.".
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/3 GET https://localhost:7007/ - - - 200 - - 19777.8997ms
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[12]
      Stream id "0HMK61V4ICAUC:00000003" write aborted by peer with error code 256.
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[10]
      Stream id "0HMK61V4ICAUC:00000003" shutting down writes because: "Connection aborted by peer (256).".
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[5]
      Connection id "0HMK61V4ICAUC" aborted by peer with error code 256.
      System.Net.Quic.QuicConnectionAbortedException: Connection aborted by peer (256).
         at System.Net.Quic.Implementations.MsQuic.MsQuicConnection.AcceptStreamAsync(CancellationToken cancellationToken)
         at System.Net.Quic.QuicConnection.AcceptStreamAsync(CancellationToken cancellationToken)
         at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Internal.QuicConnectionContext.AcceptAsync(CancellationToken cancellationToken)
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[11]
      Stream id "0HMK61V4ICAUC:00000002" read aborted by peer with error code 256.
dbug: Microsoft.AspNetCore.Server.Kestrel.Http3[45]
      Trace id "0HMK61V4ICAUC": HTTP/3 stream error "H3_NO_ERROR". An abort is being sent to the stream.
      Microsoft.AspNetCore.Connections.ConnectionAbortedException: The client closed the connection.
       ---> Microsoft.AspNetCore.Connections.ConnectionResetException: Connection aborted by peer (256).
       ---> System.Net.Quic.QuicConnectionAbortedException: Connection aborted by peer (256).
         at System.Net.Quic.Implementations.MsQuic.MsQuicConnection.AcceptStreamAsync(CancellationToken cancellationToken)
         at System.Net.Quic.QuicConnection.AcceptStreamAsync(CancellationToken cancellationToken)
         at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Internal.QuicConnectionContext.AcceptAsync(CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Internal.QuicConnectionContext.AcceptAsync(CancellationToken cancellationToken)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3.Http3Connection.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
         --- End of inner exception stack trace ---
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[6]
      Connection id "0HMK61V4ICAUC" aborted by application with error code 256 because: "The client closed the connection.".
dbug: Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[13]
      Stream id "0HMK61V4ICAUC:00000002" aborted by application with error code 256 because: "The client closed the connection.".
dbug: Microsoft.AspNetCore.Server.Kestrel.Http3[44]
      Connection id "0HMK61V4ICAUC" is closed. The last processed stream ID was 0.
dbug: Microsoft.AspNetCore.Server.Kestrel.Connections[2]
      Connection id "0HMK61V4ICAUC" stopped.
dbug: Microsoft.Extensions.Hosti...
```

When you asked about testing is this the test you were looking for (particularly with YARP) or you would be fine with any client?

Do you have any client on your mind that could be used for this test? (Other than writing some code that mimics the Integration test but instead of using in-memory connection, using a real socket)

ladeak replied:
A new test is added to src/Servers/Kestrel/test/Interop.FunctionalTests/Http3/Http3ConnectionTests.cs

halter73 replied:
> The ASP.NET Core webapi awaits Task.Delay(...) before writing the response. Until this I shut down the YARP proxy.

> ...

> EDIT: or you were just looking for an integration test?

While HttpClient doesn't currently send any GOAWAY frames for non-idle connections ongoing requests yet, we could still test the feature for idle connections when I suspect HttpClient does send GOAWAYs. So we shouldn't need any Task.Delay(...) in our end-to-end test with HttpClient.

Can we set `SocketsHttpHandler.PooledConnectionIdleTimeout` to something small and write a more end-to-end test more along the lines of one of our [Quic WebHostTests](https://github.com/dotnet/aspnetcore/blob/e971d938793f2d0b8bc96e5f839e61066d19ecea/src/Servers/Kestrel/Transport.Quic/test/WebHostTests.cs#L56)?

> A new test is added to src/Servers/Kestrel/test/Interop.FunctionalTests/Http3/Http3ConnectionTests.cs

👍 I like this too. Thanks!

ladeak replied:
> The ASP.NET Core webapi awaits Task.Delay(...) before writing the response. Until this I shut down the YARP proxy. 

I was wondering what I meant to say here. It was a 'manual' test out of curiosity hence I needed the Task.Delay. Hopefully no need for in automated tests.

As suggested, I am adding tests using `SocketsHttpHandler.PooledConnectionIdleTimeout` in Quic WebHostTests.

ladeak replied:
@halter73 I am a little puzzled: I have prepared a test as suggested with `socketsHttpHandler.PooledConnectionIdleTimeout` but I don't seem to receive any GoAway frame. I am reviewing [Http3Connection](https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/Http3Connection.cs,e0c4de20c64e0773), [Http3RequestStream](https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs,2f918e96f263c918) and [HttpConnectionPool](https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs,9833cd617ee6fa94) but I am afraid I did not find the place where the GoAway frame would be sent on idle connections. Could you help me point me to the right direction, so that I can better understand this case and prepare the test?

ladeak replied:
@halter73, could you point me to the right direction on this one?

halter73 reviewed (commented):
@Tratcher @JamesNK do you think this is useful enough for YARP and/or gRPC to add to .NET 7?

Tratcher commented:
I don't think YARP could take advantage of this since it can't (and wouldn't want to) proxy the GoAway to the destination server. YARP would be limited to canceling individual requests/streams which would not be graceful.

JamesNK commented:
> @Tratcher @JamesNK do you think this is useful enough for YARP and/or gRPC to add to .NET 7?

Is this for the situation where a client has sent a GOAWAY, but the socket is still alive, and it prevents Kestrel from shutting down without hitting a timeout?

I haven't heard anyone bring that up as a gRPC issue.

ladeak commented:
> > @Tratcher @JamesNK do you think this is useful enough for YARP and/or gRPC to add to .NET 7?
> 
> I haven't heard anyone bring that up as a gRPC issue.

If you come to the conclusion that maybe then this change should be reconsidered, I was aware that the conversation for the issue still had this open question. Even in this case, I don't mind spending the time on the implementation.

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:
```diff
@@ -968,6 +968,7 @@ private Task ProcessGoAwayFrameAsync()
         }
 
         StopProcessingNextRequest(serverInitiated: false);
```
```suggestion
        // StopProcessingNextRequest must be called before RequestClose to ensure it's considered client initiated.
        StopProcessingNextRequest(serverInitiated: false);
```

Let's add a comment to make it clear order matters here because of [the ConnectionClosedRequested registration in HttpConnection](https://github.com/dotnet/aspnetcore/blob/e971d938793f2d0b8bc96e5f839e61066d19ecea/src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs#L101).

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http3/Http3ControlStream.cs:
```diff
@@ -362,6 +362,8 @@ private void ProcessSetting(long id, long value)
     private ValueTask ProcessGoAwayFrameAsync()
     {
         EnsureSettingsFrame(Http3FrameType.GoAway);
+        _context.Connection.StopProcessingNextRequest(serverInitiated: false);
```
```suggestion
        // StopProcessingNextRequest must be called before RequestClose to ensure it's considered client initiated.
        _context.Connection.StopProcessingNextRequest(serverInitiated: false);
```

Also, not related to this PR exactly, but the two StopProcessingNextRequest methods in Http3ControlStream itself are not called by anything, right? At first, I thought you were calling one of those, and it seemed odd. Can we delete them?

ladeak replied:
Sure, I don't see those being used.

ladeak commented:
@halter73 I am a little puzzled: I have prepared a test as suggested with socketsHttpHandler.PooledConnectionIdleTimeout but I don't seem to receive any GoAway frame. I am reviewing [Http3Connection](https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/Http3Connection.cs,e0c4de20c64e0773), [Http3RequestStream](https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs,2f918e96f263c918) and [HttpConnectionPool](https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs,9833cd617ee6fa94) but I am afraid I did not find the place where the GoAway frame would be sent on idle connections. Could you help me point me to the right direction, so that I can better understand this case and prepare the test?

halter73 commented:
@ladeak Sorry for telling you to go ahead with this and then asking for these tests before more closely looking at the client behavior. I probably shouldn't have even marked it up for grabs until I had done that. I think you're right that SocketsHttpHandler simply never initiates a graceful close of the connection with a GOAWAY frame.

The client is allowed to gracefully close an HTTP/3 connection with a GOAWAY frame according to the HTTP/3 spec, but it certainly isn't required to. The server "SHOULD" send such a frame, but it's not even required to. In the HTTP/2 spec, I think the "SHOULD" goes both for both the server and the client.

@dotnet/ncl Can someone tell me off the top of their head if SocketsHttpHandler ever initiates a graceful GOAWAY for either HTTP/2 or HTTP/3 connections. If so, under what circumstances? Idle timeouts? Disposing the HttpClient with in flight requests?

MihaZupan commented:
As far as I know, `SocketsHttpHandler` will never send a `GOAWAY` on either HTTP/2 or HTTP/3.
We just close the underlying transport when we're done (disposed && no in flight requests).

halter73 commented:
Cool. I still don't see the harm in doing this. The spec allows it, and we have the tests where we manually send the GOAWAY frames. I see no reason not to fire a `ConnectionClosedRequested` in the event a GOAWAY is received. @JamesNK does this make sense to you?

MihaZupan commented:
Is it possible/expected that the interface may be notified multiple times for the same connection?

halter73 commented:
> Is it possible/expected that the interface may be notified multiple times for the same connection?

No, but it's backed by a CancellationToken, so we should be good.

https://github.com/dotnet/aspnetcore/blob/c7a4ec8c64a49f82369e85bca0b386cfa01fc207/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelConnection.cs#L18

JamesNK commented:
I think the non-test changes are fine.

Is an interop test for HTTP/3 necessary? Because we haven't built a fully-fledged HTTP/3 client for our tests you've needed to hack in parsing H3 frames and QPack in the interop tests. I think the in-memory functional test `GOAWAY_TriggersLifetimeNotification_ConnectionClosedRequested` is enough.

(we probably should at some point invest in making the in-memory methods support real HTTP/3 over QUIC, but that's a whole other thing)

ladeak commented:
I have personally felt much better after having the interop test written - to really see this in action through a socket. (but I also see your point)

@JamesNK and @halter73 , let me know how you decide, and I will remove the test if needed :)

halter73 commented:
> let me know how you decide, and I will remove the test if needed :)

I agree with @JamesNK. I think GOAWAY_TriggersLifetimeNotification_ConnectionClosedRequested` is sufficient. I would have liked an HttpClient-based test for this if it were possible, but it not. Let's remove the interop test.

ladeak committed:
Triggering Connection Lifetime Notification Feature's cancellation on GoAway

On GoAway frames triggering `IConnectionLifetimeNotificationFeature`'s cancellation for gRPC and SignalR to initiate teardown of long running requests.

## Description

For Kestrel h2 and h3 triggering `IConnectionLifetimeNotificationFeature` feature's *RequestClose()* method on GoAway frames, so that frameworks like SignalR or gRPC may proactively start closing long running request.
h3 control stream (which is the only stream that handles GoAway) also calls *StopProcessingNextRequest()* method otherwise `HttpConnection` (subscribed to `IConnectionLifetimeNotificationFeature`) would send GoAway's from server. h2 already does the same.

Adding new tests: `Http3InMemory` now implements *RequestClose()* method and `Http2TestBase` mimics sharing the same feature collection on ConnectionFeatures and by the Connection's Features properties.

Fixes #42057

JamesNK reviewed (approved):
LGTM

ladeak commented:
Could someone trigger a new build here?

JamesNK commented:
/azp run

JamesNK commented:
/azp run
